token/program-2022/src/extension/confidential_mint_burn/processor.rs

```diff
@@ -0,0 +1,428 @@
+        program_id,
+        &authority,
+        authority_info,
+        authority_info_data_len,
+        account_info_iter.as_slice(),
+    )?;
+
+    mint_burn_extension.supply_elgamal_pubkey = Some(proof_context.second_pubkey).try_into()?;
+    mint_burn_extension.confidential_supply = proof_context.second_ciphertext;
+
+    Ok(())
+}
+
+/// Processes an [UpdateAuthority] instruction.
+fn process_update_decryptable_supply(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    new_decryptable_supply: PodAeCiphertext,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let mint_info = next_account_info(account_info_iter)?;
+    let authority_info = next_account_info(account_info_iter)?;
+    let authority_info_data_len = authority_info.data_len();
+
+    check_program_account(mint_info.owner)?;
+    let mint_data = &mut mint_info.data.borrow_mut();
+    let mut mint = PodStateWithExtensionsMut::<PodMint>::unpack(mint_data)?;
+    let mint_burn_extension = mint.get_extension_mut::<ConfidentialMintBurn>()?;
+
+    let authority = Option::<Pubkey>::from(mint_burn_extension.authority)
+        .ok_or(TokenError::NoAuthorityExists)?;
+
+    Processor::validate_owner(
+        program_id,
+        &authority,
+        authority_info,
+        authority_info_data_len,
+        account_info_iter.as_slice(),
+    )?;
+
+    mint_burn_extension.decryptable_supply = new_decryptable_supply;
+
+    Ok(())
+}
+
+/// Processes a [ConfidentialMint] instruction.
+#[cfg(feature = "zk-ops")]
+fn process_confidential_mint(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: &MintInstructionData,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+    let token_account_info = next_account_info(account_info_iter)?;
+    let mint_info = next_account_info(account_info_iter)?;
+
+    check_program_account(mint_info.owner)?;
+    let mint_data = &mut mint_info.data.borrow_mut();
+    let mut mint = PodStateWithExtensionsMut::<PodMint>::unpack(mint_data)?;
+
+    let auditor_elgamal_pubkey = mint
+        .get_extension::<ConfidentialTransferMint>()?
+        .auditor_elgamal_pubkey;
+    let mint_burn_extension = mint.get_extension_mut::<ConfidentialMintBurn>()?;
+
+    let proof_context = verify_mint_proof(
+        account_info_iter,
+        data.equality_proof_instruction_offset,
+        data.ciphertext_validity_proof_instruction_offset,
+        data.range_proof_instruction_offset,
+    )?;
+
+    check_program_account(token_account_info.owner)?;
+    let token_account_data = &mut token_account_info.data.borrow_mut();
+    let mut token_account = PodStateWithExtensionsMut::<PodAccount>::unpack(token_account_data)?;
+
+    let authority_info = next_account_info(account_info_iter)?;
+    let authority_info_data_len = authority_info.data_len();
+
+    let authority = Option::<Pubkey>::from(mint_burn_extension.authority)
+        .ok_or(TokenError::NoAuthorityExists)?;
```

Comment:
Currently the confidential mint burn does not disable mints of normal tokens, so they could differ. I'm not sure if there's really a use case for a token that has both a normal and confidential balance in use though. Should enabling confidential-mint-burn disable the normal mint function then?

Reply:
Confidential mint / burn does currently disable minting of normal tokens (see my other comment).

We could have both supported at once, but that seems confusing, since they would effectively be two different token types that can't interact with each other. And there would be a risk of overflowing u64::MAX for the total supply of non-confidential + confidential

Reply:
Yup seems that it does actually disable it... it's been a while since I initially wrote this code 🤦‍♂️

Reply:
Alright switched everything over to use the base mint authority